racker/salus-common
# Added a conditionalized RestTemplate customizer for SSL

# Resolves

https://jira.rax.io/browse/SALUS-716

# What

We need to enable secure communication between the admin/public API proxy and the backend microservices.

# How

Spring Boot [makes it easy to configure SSL for the embedded server](https://docs.spring.io/spring-boot/docs/current/reference/html/howto.html#howto-configure-ssl), but leaves you hanging for the client/RestTemplate side of things. [They provide some info about customizing the RestTemplate instances](https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-resttemplate-customization) but don't provide the final bits for SSL client config. For that remaining bit I took some educated guesses at the HttpClient builder classes provided by Apache HttpClient.

# How to test

Created a CA truststore and server keystore borrowing some techniques from my UMB setup scripts and putting together this helper script:

https://gist.github.com/itzg/cffa82f6cb04af68dedf49ffc1531604#file-build-sh

Configured monitor management with the ssl properties from that Spring section mentioned above.

Configured public api with the following environment variables to confirm kubernetes configurability:

```
SALUS_RESTCLIENT_SSL_TRUSTSTORE=file:///tmp/certs/out/ca-truststore.jks
SALUS_RESTCLIENT_SSL_TRUSTSTOREPASSWORD=...
```

I also had to set the property `salus.services.monitor-management-url` to `https://localhost:8089` to ensure it used SSL.

With all that I did a simple request for `/schema/monitor-plugins`

cc @shawnashlee
